File: samples/vision/unetmobilenet/runtime/cpp/src/unetmobilenet.cpp
// Copyright (c) 2026 D-Robotics Corporation
// SPDX-License-Identifier: Apache-2.0
#include "unetmobilenet.hpp"
#include <cstring>
#include <stdexcept>
#include <utility>
#include <opencv2/imgproc.hpp>

namespace unetmobilenet {
UnetMobileNetTask::UnetMobileNetTask(RawRunner runner):runner_(std::move(runner)) {
    if(!runner_)throw std::invalid_argument("A raw runner is required");
}
PreparedInput UnetMobileNetTask::pre_process(const cv::Mat& image) const {
    if(image.empty() || image.type()!=CV_8UC3)throw std::invalid_argument("Expected nonempty BGR uint8 image");
    cv::Mat resized,i420;
    cv::resize(image,resized,cv::Size(2048,1024),0,0,cv::INTER_AREA);
    cv::cvtColor(resized,i420,cv::COLOR_BGR2YUV_I420);
    PreparedInput result{cv::Mat(1024,2048,CV_8UC1),cv::Mat(512,1024,CV_8UC2),{image.rows,image.cols}};
    const auto* source=i420.ptr<unsigned char>();
    std::memcpy(result.y.data,source,1024*2048);
    const auto* u=source+1024*2048;
    const auto* v=u+512*1024;
    auto* uv=result.uv.ptr<unsigned char>();
    for(int i=0;i<512*1024;++i) { uv[2*i]=u[i];uv[2*i+1]=v[i]; }
    return result;
}
RawScores UnetMobileNetTask::forward(const PreparedInput& prepared) const {
    return runner_(prepared.y,prepared.uv);
}
cv::Mat UnetMobileNetTask::post_process(const RawScores& raw,const ImageContext& context) const {
    const auto labels=decode_scores(raw,context.original_height,context.original_width);
    cv::Mat result(context.original_height,context.original_width,CV_32S);
    std::memcpy(result.data,labels.data(),labels.size()*sizeof(std::int32_t));
    return result;
}
cv::Mat UnetMobileNetTask::predict(const cv::Mat& image) const {
    const auto prepared=pre_process(image);
    return post_process(forward(prepared),prepared.context);
}
}  // namespace unetmobilenet

File: samples/vision/unetmobilenet/runtime/python/unetmobilenet.py
# Copyright (c) 2026 D-Robotics Corporation
# SPDX-License-Identifier: Apache-2.0
"""Cityscapes segmentation stages; file IO, rendering and SDK loading live elsewhere."""
from dataclasses import dataclass
from typing import Mapping

import cv2
import numpy as np

from samples._shared.image import bgr_to_nv12_planes
from samples._shared.quantization import dequantize_tensor
from samples.vision.unetmobilenet.runtime.python.model_binding import ModelBinding


@dataclass(frozen=True)
class ImageContext:
    original_height: int
    original_width: int


@dataclass(frozen=True)
class PreparedInput:
    tensors: Mapping[str, np.ndarray]
    context: ImageContext


class UnetMobileNetTask:
    """BGR → split NV12 → raw logits → original-resolution int32 IDs."""

    def __init__(self, runner, binding: ModelBinding):
        self.runner = runner
        self.binding = binding

    def pre_process(self, image: np.ndarray) -> PreparedInput:
        """Stretch nonempty BGR uint8 HWC with INTER_AREA to 2048×1024.

        Returns Y [1,1024,2048,1], UV [1,512,1024,2], both contiguous uint8,
        plus immutable per-call original geometry. No normalization/letterbox.
        ValueError rejects non-images, empty dimensions and non-uint8 input.
        """
        if (not isinstance(image, np.ndarray) or image.ndim != 3
                or image.shape[2] != 3 or image.dtype != np.uint8
                or not all(image.shape[:2])):
            raise ValueError('Expected nonempty BGR uint8 HWC input')
        resized = cv2.resize(image, (self.binding.input_width, self.binding.input_height),
                             interpolation=cv2.INTER_AREA)
        y, uv = bgr_to_nv12_planes(resized)
        return PreparedInput({self.binding.y_name: y, self.binding.uv_name: uv},
                             ImageContext(*image.shape[:2]))

    def forward(self, tensors: Mapping[str, np.ndarray]) -> np.ndarray:
        """Return runner-validated raw [1,H,W,19] int32/float32 unchanged."""
        return self.runner(tensors)

    def post_process(self, raw: np.ndarray, context: ImageContext) -> np.ndarray:
        """Decode scores and resize class IDs directly to the original image.

        SCALE logits are affine-dequantized only here, using float64 comparison
        to retain int32 differences; explicit NONE int32 and F32 remain raw.
        Argmax ties choose the lowest class ID. INTER_NEAREST restores geometry;
        result is an owned int32 [original_height, original_width] mask (0..18).
        Invalid logits/context raise ValueError. No coloring or blending occurs.
        """
        meta = self.binding.metadata
        name = self.binding.output_name
        if (not isinstance(raw, np.ndarray) or raw.shape != meta.output_shapes[name]
                or raw.dtype != np.dtype(meta.output_dtypes[name]) or not np.isfinite(raw).all()):
            raise ValueError('Logits do not match the bound shape/dtype or contain nonfinite values')
        if (not isinstance(context, ImageContext) or context.original_height <= 0
                or context.original_width <= 0):
            raise ValueError('A valid per-call ImageContext is required')
        scores = raw
        if raw.dtype == np.int32:
            scores = dequantize_tensor(raw, meta.output_quants[name], dtype='float64')
        labels = np.argmax(scores[0], axis=-1).astype(np.int32)
        return cv2.resize(labels, (context.original_width, context.original_height),
                          interpolation=cv2.INTER_NEAREST).copy()

    def predict(self, image: np.ndarray) -> np.ndarray:
        """Run the same three stages and return class IDs, not a visualization."""
        prepared = self.pre_process(image)
        return self.post_process(self.forward(prepared.tensors), prepared.context)
